ingestion/functions/aws_access/EventBridgeClient.py

```diff
@@ -1,5 +1,114 @@
+    # add 7 if negative
+    normalised_difference = weekday_difference if weekday_difference >= 0 else weekday_difference + 7
+    # add one, as days of month are counted from 1
+    first_requested_weekday = normalised_difference + 1
+    # add 7*(week_of_month - 1) days
+    return first_requested_weekday + (7 * (week_of_month - 1))
+
+
+class ScheduleRule:
+    """Representation of an EventBridge schedule rule"""
+    def __init__(self, rule_description: dict):
+        self.rule_description = rule_description
+    
+    def schedule(self):
+        return self.rule_description.get("ScheduleExpression")
+    
+    def is_ingestion_rule(self):
+        return self.description().startswith("Scheduled Batch ingestion rule")
+    
+    def description(self):
+        return self.rule_description.get("Description", "")
+
+    def ingestion_source_id(self):
+        assert self.is_ingestion_rule()
+        return re.search("[a-z0-9]{24}", self.description()).group(0)
+
+    def date_of_requested_weekday_in_month(self, weekday, week_of_month, year, month):
+        return date_of_requested_weekday_in_month(weekday, week_of_month, year, month)
+
+    def oldest_file_age(self, now):
+        """
+        What is the age in days of the oldest source file for an ingestion rule that we can keep?
+        now: reference time from which to compute file age
+        """
+        assert self.is_ingestion_rule()
+        schedule = self.schedule()
+        if schedule.startswith('rate'):
+            components = re.match(r'rate\(([0-9]+) ([a-z]+)\)', schedule)
+            assert components is not None # I'm assuming the format is rate(1 day)
+            assert components.group(2).startswith('day') # I'm assuming the unit is day or days
+            number_of_days = int(components.group(1))
+            assert number_of_days > 0 # I'm assuming a positive schedule
+            age = timedelta(days = -2 * number_of_days - GRACE_PERIOD_IN_DAYS)
+        else:
+            assert schedule.startswith('cron') # I'm assuming that if it isn't a rate, it's a cron
+            components = re.match(r'cron\(([0-9]+) ([0-9]+) (\?|\*) (\*) ([0-6]#[1-4]|[A-Z]{3}|[0-6]|\?) (\*)\)', schedule)
+            assert components is not None # Assuming a restricted subset of https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-create-rule-schedule.html rules
+            # I also assume that you never ask for the fifth weekday of the month, because that's a very odd schedule
+            (minute, hour, day_of_month, month, day_of_week, year) = components.groups()
+            # calculate from midnight as we don't want odd seconds etc
+            today = datetime(now.year, now.month, now.day)
+            our_day = today.isoweekday() % 7 # ISO runs 1-7 Mon-Sun, we want 0-6 Sun-Sat
+            # Because of my assumptions above in the regex, I only need the day of the week here
+            # When we are using Python 3.10+, this should be a match
```

Comment:
Is this using a feature only in 3.10?

Reply:
The intent of the comment is that I didn't use a 3.10-only feature, but it would be better if I could (all those if/elif/else can be a match/case/case which is a nicer way of expressing this).